### DISCUSSION TITLE
Integration of Loco Positioning System with external compute boards

### DISCUSSION THREAD

**Discussion by deepparikh** [bitcraze/discussions#775] [Q&A] [answered]
Hello,

I am considering the Loco positioning system for proximity operations of CubeSats. We have our own compute board (Consider Teensy 4.1) and an IMU. The computer board handles the data collection from IMU and has EKF based AHRS. For the next phase, we will be working on combining UWB based range measurements and IMU data to estimate 6DOF pose (position and attitude) of moving satellite modules with respect to a fixed satellite (with anchors). I am primarily curious about the following,

1. What is the distribution of the computation between the node and the Crazyflie 2.1 control board? I suppose the fusion of processed range measurements and IMU data happens at the control board, but I am curious how much of signal processing is done by the node.  

2. Can I use the Loco Positioning Node directly with external compute boards (like Teensy 4.1) ?  Since we are planning to have our own estimation algorithm and IMU, in ideal scenario, we still want node to do heavy lifting of the processing of UWB range measurements and provide processed range data to external compute board. 

3. A few previous threads mention that the similar integration with pixhawk was achieved using bolt control board and Loco node. If the proposed workflow of point 2 is not possible, what additional hardware I would need? For each node, do I need bolt control board or I can still use the standard 2.1 control board ? However, this is highly undesirable workaround as the role of control board will be just to relay the range data, rendering all other avenues like IMU, radio, motor drivers etc. unused. 

Looking forward to valuable insights from the community.

**Comment by krichardsson** [comment]
This sounds like an exciting, and unusual project!

Do you plan to use Two Way Ranging or a TDoA mode? My guess is TWR, this will be the more robust solution (from a geometrical point of view) and will work well if you only need to position one CubeSat. If you want to position more than one cubeSat, you can perhaps use some sort of time slotted scheme where the cubeSats range with the fixed sat one by one.
The other option for multiple cubeSats is one of the TDoA options, but I think the geometry is very bad and I'm not sure how good the position estimate would be.

> What is the distribution of the computation between the node and the Crazyflie 2.1 control board? I suppose the fusion of processed range measurements and IMU data happens at the control board, but I am curious how much of signal processing is done by the node.

Yes, the fusion of processed range measurements and IMU data is done in the Crazyflie. I don't have any numbers for this, and it will be a guess. The work in the anchors is mainly shuffling some data around and does not require that much computations (mainly integers). The Crazyflie on the other hand is more or less doing the same type of data-shuffling, but on top of that also the state estimation. The state estimation is implemented as a kalman filter which requires a some matrix operations with floats. 
A very rough guess would be that the CF is doing 5-10 times more work than the node. 

> Can I use the Loco Positioning Node directly with external compute boards (like Teensy 4.1) ? Since we are planning to have our own estimation algorithm and IMU, in ideal scenario, we still want node to do heavy lifting of the processing of UWB range measurements and provide processed range data to external compute board.

In a "normal" loco system, the nodes are not connected to a computer, except when configuring them. During operation the anchors are only connected for power.
Is the idea to get the ranging data from the anchors to enable state estimation of the CubeSat in the fixed sat? This is not supported by the standard firmware but could probably be added.
Connections to the node is usually done via USB, but there is also a UART that could be used.

> A few previous threads mention that the similar integration with pixhawk was achieved using bolt control board and Loco node. If the proposed workflow of point 2 is not possible, what additional hardware I would need? For each node, do I need bolt control board or I can still use the standard 2.1 control board ? However, this is highly undesirable workaround as the role of control board will be just to relay the range data, rendering all other avenues like IMU, radio, motor drivers etc. unused.

I'm a bit fuzzy on exactly how you want to implement your system (where to do the state estimation and type of ranging) which makes it a bit hard to suggest solutions. Would it be possible to clarify this a bit to align my brain with your view of the problem?

There is no significant difference between the Bolt and CF2.1 when it comes to the Loco deck, CPU and processing power. The difference is mainly related to the motor drivers.

Just for information, it is also possible to add more data to the UWB messages if you want to use this channel to send data between the fixed sat and CubeSat.

**Reply by deepparikh** [reply]
Thank you so much @krichardsson for your comprehensive response and apologies for not including sufficient details. I have a tendency develop a tunnel vision while working on a project for a long time. Please find my following response which I hope will add more clarity.

`Do you plan to use Two Way Ranging or a TDoA mode?`
My initial plan is to work with two CubeSats and make them dock with each other.  I hope to start with TWR but gradually move to TDOA3 as there will be addition of more CubeSats up to maximum 10. But if we realize in the process that TDOA3 is not providing sufficient range accuracy, I am also open to alternate options which might increase accuracy at the cost of update rate. 

`Is the idea to get the ranging data from the anchors to enable state estimation of the CubeSat in the fixed sat?`
No, my application is exactly same as the CF. The node will provide range data to moving CubeSat and the individual CubeSats need to be aware of their relative positon. The control scheme is decentralized, so each agent knows their task. Please refer to the attached images for more details. 

`Would it be possible to clarify this a bit to align my brain with your view of the problem?`

![life_line_diagram](https://github.com/bitcraze/discussions/assets/5811139/54aac0f2-50bd-4d78-995d-c793997bd050)
Sure. As per the attached life line diagram, I suppose during normal operation, the CF2.1/bolt flight controller pings for a range data and the LPS node responds with an appropriate message. My initial goal is to explore feasibility of 'tricking' LPS node by still using the similar communication interface from Teensy 4.1. Can we implement a communication subroutine in Teensy 4.1 (or any other boards) that 'mimics' data collection behavior of CF2.1/bolt such that LPS node always thinks that it is connected with CF2.1/bolt instead of Teensy 4.1? This seems like a least effort solution to get the range data from LPS node, but if there are other efficient avenues, I would be glad to try that.

![block_diagram](https://github.com/bitcraze/discussions/assets/5811139/c1fe0708-f623-4d5f-b699-109e70d627a2)
Coming to the pose estimation, as per the attached block diagram, the Teensy 4.1 , IMU and LPS node are mounted on a moving CubeSat. The teensy collects raw data from IMU and provides attitude estimate. The intended workflow is get the range data from LPS node and then combine IMU and UWB range data to estimate relative position of CubeSat and orientation onboard the moving CubeSata.

Our research group focuses on novel sensor technologies and estimation algorithms, hence we will be validating performance of various estimators along the process of finding the best one which meets our requirement. 

`Just for information, it is also possible to add more data to the UWB messages if you want to use this channel to send data between the fixed sat and CubeSat.` 
This seems a very lucrative idea. While we do have a telemetry link between CubeSats and fixed satellite, we can consider replacing it with UWB messages if the tradeoff in range update is not significant.

**Reply by krichardsson** [reply, marked as answer]
Thanks for the clarification!

I actually think there is a solution that is fairly simple to implement and that will scale to 10 cube sats. 

The idea is to only use Loco Nodes (no need for a Crazyflie or Bolt) and run them in TDoA3 mode. The key here is that in TDoA3, the Loco nodes are actually doing [TWR with each other](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/functional-areas/tdoa_principles/) and the only functionality that is missing is to get the ranging data out of the node on the cube sats. 
TDoA3 was designed to handle systems with a dynamic number of nodes and as far as I can see it should work pretty much out of the box, with a few question marks to check (see below). As an extra bonus the cube stats will also get the distance to the other cube sats, not only the anchors on the fixed sat.

The missing functionality in the nodes is to get the ranging data to your host (Teensy), but I think it should be as easy as simply printing the ranging values to the serial port/USB when they are computed. We have a similar functionality for TWR when the node is configured as a Tag, but it is not implemented for TDoA3. I think it would be a nice addition and could be a configuration in the menu (please consider contributing!)

## Concerns and question marks

The default settings in the anchors for TDoA3 keeps track of the distance to 16 nodes. If you need more nodes than this you might want to increase this number.
When I wrote the TDoA3 code for the nodes I assumed that anchors are not moving around (as much as a Crazyflie), and there might be some optimizations (like timeouts) that are chosen based on this thinking. In your case, the nodes in the system will be moving, but I assume that it will be very slowly and I don't think it will be a problem.
The CPU on the Loco Nodes is pretty simple and I remember that TDoA3 was close to using both the available RAM and CPU power, but I think that there will be enough resources left to print the data to the serial port. 

## Configuration

You may want to consider the [Long Range mode](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/user-guides/tdoa3_long_range/) for TDoA3 to reach further? I'm not sure what your requirements are.


When I designed TDoA3 5 years ago, this was not the use case I had in mind, but it is super exciting to use it as a dynamic Two Way Ranging system instead! If you need more information, pointers to the code or similar, just let me know!

**Reply by deepparikh** [reply]
Thank you so much for your optimism and for offering to provide your valuable insights. 

`The key here is that in TDoA3, the Loco nodes are actually doing TWR with each other`
This was something I had in mind as an enhancement, but it is comforting to know that it has already been implemented. 

`I think it would be a nice addition and could be a configuration in the menu (please consider contributing!)`
Absolutely ! 

`You may want to consider the` [Long Range mode](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/user-guides/tdoa3_long_range/)` for TDoA3 to reach further?`
Yes, we are considering the proximity operations in the range of 50 m to docking. Hence the Long Range mode should be the correct avenue. But we are flexible on the maximum range aspect. As long as it is more than 10 m, we should make that work. 

Although we have a few unknowns, this seems a worthy exploration. I am proceeding with the procurement of [Loco explorer bundle](https://store.bitcraze.io/products/indoor-explorer-bundle), additional [Loco positioning deck](https://store.bitcraze.io/products//loco-positioning-deck) and a couple of extra batteries for the start. @krichardsson I will find a direct communication channel to reach out for the further advancements.
